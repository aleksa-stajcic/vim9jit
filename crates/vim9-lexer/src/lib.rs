#![allow(dead_code)]

use std::collections::VecDeque;
use std::fmt::Debug;

#[derive(Clone, PartialEq)]
pub struct Span {
    start_row: usize,
    start_col: usize,
    end_row: usize,
    end_col: usize,
}

impl Debug for Span {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "({},{})->({},{})",
            self.start_row, self.start_col, self.end_row, self.end_col
        )
    }
}

#[derive(Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub text: String,
    pub span: Span,
}

impl Debug for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Token({:?}, {:?}, {:?})", self.kind, self.text, self.span)
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    Illegal,
    EndOfFile,

    // TODO: Is this crazy for this lang??
    EndOfLine,
    Comment,

    // Identifiers and literals
    Identifier,
    Integer,

    // Reserved Tokens
    True,
    False,
    Null,

    // Operators
    Equal,
    Plus,
    Minus,
    Mul,
    Div,
    PlusEquals,
    MinusEquals,
    MulEquals,
    DivEquals,

    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    EqualTo,
    NotEqualTo,
    Or,
    And,

    // Delimiters
    Comma,
    Colon,

    SingleQuoteString,
    DoubleQuoteString,

    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
}

pub struct Lexer {
    pub input: String,
    pub position: usize,
    pub read_position: usize,
    pub ch: Option<char>,

    chars: Vec<char>,
}

impl Debug for Lexer {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Lexer {{ position: {}, read_position: {}, ch: {:?} }}",
            self.position, self.read_position, self.ch
        )
    }
}

impl Lexer {
    fn new(input: &str) -> Self {
        let input = input.to_string();
        let chars = input.clone().chars().collect();
        Self {
            input,
            position: 0,
            read_position: 0,
            ch: None,
            chars,
        }
    }

    fn make_span(&self, start: usize, end: usize) -> Span {
        // TODO: This is broken for doing stuff with empty strings... :'(
        assert!(
            start <= end,
            "start must be less than end. {} {}\n{:#?}",
            start,
            end,
            self
        );

        let mut span = Span {
            start_row: 0,
            start_col: 0,
            end_row: 0,
            end_col: 0,
        };

        let mut row = 0;
        let mut row_start = 0;
        for (idx, ch) in self.input.chars().enumerate() {
            if idx == start {
                span.start_row = row;
                span.start_col = idx - row_start;
            }

            if idx == end {
                span.end_row = row;
                span.end_col = idx - row_start;
                return span;
            }

            if ch == '\n' {
                row += 1;
                row_start = idx + 1;
            }
        }

        unreachable!(
            "Should always fine the start and end for a span:\n{} {} {}",
            start,
            end,
            self.input.chars().count()
        )
    }

    pub fn read_char(&mut self) {
        match self.chars.get(self.read_position) {
            Some(&c) => {
                self.ch = Some(c);
                self.position = self.read_position;
                self.read_position += 1;
            }
            None => {
                self.ch = None;
            }
        }
    }

    fn read_number(&mut self) -> Token {
        let position = self.position;
        while let Some(ch) = self.ch && ch.is_numeric(){
            self.read_char();
        }

        Token {
            kind: TokenKind::Integer,
            text: self.chars[position..self.position].iter().collect(),
            span: self.make_span(position, self.position),
        }
    }

    fn read_until(&mut self, until: char, kind: TokenKind) -> Token {
        self.read_char();
        if let Some(ch) = self.ch && ch == until {
            return Token {
                kind,
                text: self.chars[self.position..self.position].iter().collect(),
                span: self.make_span(self.position, self.position ),
            };
        }

        let position = self.position;

        while let Some(ch) = self.ch && ch != until {
            self.read_char();
        }

        Token {
            kind,
            text: self.chars[position..self.position].iter().collect(),
            span: self.make_span(position, self.position - 1),
        }
    }

    fn read_comment(&mut self) -> Token {
        let position = self.position;
        while let Some(ch) = self.ch && ch != '\n' {
            self.read_char();
        }

        Token {
            kind: TokenKind::Comment,
            text: self.chars[position..self.position].iter().collect(),
            span: self.make_span(position, self.position),
        }
    }

    fn read_identifier(&mut self) -> Token {
        let position = self.position;
        while let Some(ch) = self.ch && is_identifier(ch) {
            self.read_char();
        }

        let text: String = self.chars[position..self.position].iter().collect();
        let kind = match text.as_str() {
            "true" => TokenKind::True,
            "false" => TokenKind::False,
            "null" => TokenKind::Null,
            _ => TokenKind::Identifier,
        };

        Token {
            kind,
            text,
            span: self.make_span(position, self.position),
        }
    }

    fn skip_whitespace(&mut self) {
        if self.ch == Some('\n') {
            return;
        }

        while let Some(ch) = self.ch && ch.is_ascii_whitespace() {
            if ch == '\n' {
                return
            }

            self.read_char();
        }
    }

    fn peek_char(&mut self) -> Option<&char> {
        self.chars.get(self.position + 1)
    }

    fn if_peek(&mut self, peeked: char, no: TokenKind, yes: TokenKind) -> Token {
        if let Some(ch) = self.peek_char() {
            if *ch == peeked {
                let position = self.position;
                self.read_char();

                Token {
                    kind: yes,
                    text: self.chars[position..=self.position].iter().collect(),
                    span: self.make_span(position, self.position),
                }
            } else {
                Token {
                    kind: no,
                    text: self.ch.unwrap().to_string(),
                    span: self.make_span(self.position, self.position),
                }
            }
        } else {
            Token {
                kind: TokenKind::EndOfFile,
                text: "".to_string(),
                span: self.make_span(self.position, self.position),
            }
        }
    }

    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();

        let tok = match self.ch {
            Some(ch) => {
                macro_rules! literal {
                    ($kind:tt) => {
                        Token {
                            kind: TokenKind::$kind,
                            text: ch.to_string(),
                            span: self.make_span(self.position, self.position),
                        }
                    };
                }

                match ch {
                    // Operators that can optionally have an additional equals
                    '=' => self.if_peek('=', TokenKind::Equal, TokenKind::EqualTo),
                    '+' => self.if_peek('=', TokenKind::Plus, TokenKind::PlusEquals),
                    '-' => self.if_peek('=', TokenKind::Minus, TokenKind::MinusEquals),
                    '*' => self.if_peek('=', TokenKind::Mul, TokenKind::MulEquals),
                    '/' => self.if_peek('=', TokenKind::Div, TokenKind::DivEquals),
                    '>' => self.if_peek('=', TokenKind::GreaterThan, TokenKind::GreaterThanOrEqual),
                    '<' => self.if_peek('=', TokenKind::LessThan, TokenKind::LessThanOrEqual),
                    '|' => self.if_peek('|', TokenKind::Illegal, TokenKind::Or),
                    '&' => self.if_peek('&', TokenKind::Illegal, TokenKind::And),

                    ':' => literal!(Colon),
                    '(' => literal!(LeftParen),
                    ')' => literal!(RightParen),
                    '[' => literal!(LeftBracket),
                    ']' => literal!(RightBracket),
                    '{' => literal!(LeftBrace),
                    '}' => literal!(RightBrace),
                    '\n' => literal!(EndOfLine),
                    '#' => self.read_comment(),

                    // TODO: Handle escaped strings.
                    '\'' => self.read_until('\'', TokenKind::SingleQuoteString),
                    '"' => self.read_until('"', TokenKind::DoubleQuoteString),

                    _ => {
                        // Token
                        if is_letter(ch) {
                            return self.read_identifier();
                        } else if ch.is_ascii_digit() {
                            return self.read_number();
                        } else {
                            Token {
                                kind: TokenKind::Illegal,
                                text: ch.to_string(),
                                span: self.make_span(self.position, self.position),
                            }
                        }
                    }
                }
            }
            None => Token {
                kind: TokenKind::EndOfFile,
                text: "".to_string(),
                span: self.make_span(self.position, self.position),
            },
        };

        self.read_char();
        tok
    }
}

fn is_identifier(ch: char) -> bool {
    ch.is_alphanumeric() || ch == '_'
}

fn is_letter(ch: char) -> bool {
    'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z'
}

fn is_digit(ch: char) -> bool {
    '0' <= ch && ch <= '9'
}

pub fn new_lexer(input: &str) -> Lexer {
    let mut l = Lexer::new(input);
    l.read_char();
    l
}

pub fn snapshot_lexing(input: &str) -> String {
    let mut lexer = new_lexer(input);

    let mut tokens = VecDeque::new();
    loop {
        let tok = lexer.next_token();
        if tok.kind == TokenKind::EndOfFile {
            break;
        }

        tokens.push_back(tok);
    }

    let mut output = String::new();
    for (row, line) in input.lines().enumerate() {
        output += line;
        output += "\n";

        while let Some(tok) = tokens.pop_front() {
            if tok.span.start_row != tok.span.end_row {
                panic!("We haven't handled this yet");
            }

            if tok.span.start_row != row {
                tokens.push_front(tok);
                break;
            }

            output += &" ".repeat(tok.span.start_col);
            output += &"^".repeat(tok.span.end_col + 1 - tok.span.start_col);
            output += &format!(" {:?}", tok);
            output += "\n"
        }
    }

    output
}

#[cfg(test)]
mod test {
    use super::*;

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_lexing(contents));
                });
            }
        };
    }

    snapshot!(test_lexer_1, "../testdata/snapshots/lexer_1.vim");
    snapshot!(test_comparisons, "../testdata/snapshots/comparisons.vim");
    snapshot!(test_string, "../testdata/snapshots/string.vim");

    // TODO: Check more thoroughly
    snapshot!(test_matchparen, "../testdata/snapshots/matchparen.vim");
}
