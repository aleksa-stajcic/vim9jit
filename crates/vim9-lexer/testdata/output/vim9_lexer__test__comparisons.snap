---
source: crates/vim9-lexer/src/lib.rs
assertion_line: 738
expression: snapshot_lexing(contents)
---
vim9script
^^^^^^^^^^ Token(Identifier, "vim9script", (0,0)->(0,10))
           Token(EndOfLine, "\n", (0,10)->(0,10))

 Token(EndOfLine, "\n", (1,0)->(1,0))
var x = 10
^^^ Token(Identifier, "var", (2,0)->(2,3))
    ^ Token(Identifier, "x", (2,4)->(2,5))
      ^ Token(Equal, "=", (2,6)->(2,7))
        ^^ Token(Integer, "10", (2,8)->(2,10))
           Token(EndOfLine, "\n", (2,10)->(2,10))
if x == 10 || x >= 10 || x <= 10 || x > 10 || x < 10
^^ Token(Identifier, "if", (3,0)->(3,2))
   ^ Token(Identifier, "x", (3,3)->(3,4))
     ^^ Token(EqualTo, "==", (3,5)->(3,7))
        ^^ Token(Integer, "10", (3,8)->(3,10))
           ^^ Token(Or, "||", (3,11)->(3,13))
              ^ Token(Identifier, "x", (3,14)->(3,15))
                ^^ Token(GreaterThanOrEqual, ">=", (3,16)->(3,18))
                   ^^ Token(Integer, "10", (3,19)->(3,21))
                      ^^ Token(Or, "||", (3,22)->(3,24))
                         ^ Token(Identifier, "x", (3,25)->(3,26))
                           ^^ Token(LessThanOrEqual, "<=", (3,27)->(3,29))
                              ^^ Token(Integer, "10", (3,30)->(3,32))
                                 ^^ Token(Or, "||", (3,33)->(3,35))
                                    ^ Token(Identifier, "x", (3,36)->(3,37))
                                      ^ Token(GreaterThan, ">", (3,38)->(3,39))
                                        ^^ Token(Integer, "10", (3,40)->(3,42))
                                           ^^ Token(Or, "||", (3,43)->(3,45))
                                              ^ Token(Identifier, "x", (3,46)->(3,47))
                                                ^ Token(LessThan, "<", (3,48)->(3,49))
                                                  ^^ Token(Integer, "10", (3,50)->(3,52))
                                                     Token(EndOfLine, "\n", (3,52)->(3,52))
  x += 1
  ^ Token(Identifier, "x", (4,2)->(4,3))
    ^^ Token(PlusEquals, "+=", (4,4)->(4,6))
       ^ Token(Integer, "1", (4,7)->(4,8))
         Token(EndOfLine, "\n", (4,8)->(4,8))
  x *= 1
  ^ Token(Identifier, "x", (5,2)->(5,3))
    ^^ Token(MulEquals, "*=", (5,4)->(5,6))
       ^ Token(Integer, "1", (5,7)->(5,8))
         Token(EndOfLine, "\n", (5,8)->(5,8))
  x /= 1
  ^ Token(Identifier, "x", (6,2)->(6,3))
    ^^ Token(DivEquals, "/=", (6,4)->(6,6))
       ^ Token(Integer, "1", (6,7)->(6,8))
         Token(EndOfLine, "\n", (6,8)->(6,8))
  x -= 1
  ^ Token(Identifier, "x", (7,2)->(7,3))
    ^^ Token(MinusEquals, "-=", (7,4)->(7,6))
       ^ Token(Integer, "1", (7,7)->(7,8))
         Token(EndOfLine, "\n", (7,8)->(7,8))
  echo x
  ^^^^ Token(Identifier, "echo", (8,2)->(8,6))
       ^ Token(Identifier, "x", (8,7)->(8,8))
         Token(EndOfLine, "\n", (8,8)->(8,8))
endif
^^^^^ Token(Identifier, "endif", (9,0)->(9,5))
      Token(EndOfLine, "\n", (9,5)->(9,5))

 Token(EndOfLine, "\n", (10,0)->(10,0))
var a = "hello"
^^^ Token(Identifier, "var", (11,0)->(11,3))
    ^ Token(Identifier, "a", (11,4)->(11,5))
      ^ Token(Equal, "=", (11,6)->(11,7))
         ^^^^ Token(DoubleQuoteString, "hello", (11,9)->(11,13))
                Token(EndOfLine, "\n", (11,15)->(11,15))
var b = "world"
^^^ Token(Identifier, "var", (12,0)->(12,3))
    ^ Token(Identifier, "b", (12,4)->(12,5))
      ^ Token(Equal, "=", (12,6)->(12,7))
         ^^^^ Token(DoubleQuoteString, "world", (12,9)->(12,13))
                Token(EndOfLine, "\n", (12,15)->(12,15))

 Token(EndOfLine, "\n", (13,0)->(13,0))
var equal = a == b || a ==# b
^^^ Token(Identifier, "var", (14,0)->(14,3))
    ^^^^^ Token(Identifier, "equal", (14,4)->(14,9))
          ^ Token(Equal, "=", (14,10)->(14,11))
            ^ Token(Identifier, "a", (14,12)->(14,13))
              ^^ Token(EqualTo, "==", (14,14)->(14,16))
                 ^ Token(Identifier, "b", (14,17)->(14,18))
                   ^^ Token(Or, "||", (14,19)->(14,21))
                      ^ Token(Identifier, "a", (14,22)->(14,23))
                        ^^^ Token(EqualTo, "==#", (14,24)->(14,27))
                            ^ Token(Identifier, "b", (14,28)->(14,29))
                              Token(EndOfLine, "\n", (14,29)->(14,29))
var equal_ins = a ==? b
^^^ Token(Identifier, "var", (15,0)->(15,3))
    ^^^^^^^^^ Token(Identifier, "equal_ins", (15,4)->(15,13))
              ^ Token(Equal, "=", (15,14)->(15,15))
                ^ Token(Identifier, "a", (15,16)->(15,17))
                  ^^^ Token(EqualToIns, "==?", (15,18)->(15,21))
                      ^ Token(Identifier, "b", (15,22)->(15,23))
                        Token(EndOfLine, "\n", (15,23)->(15,23))

 Token(EndOfLine, "\n", (16,0)->(16,0))
var neq = a != b || a !=# b
^^^ Token(Identifier, "var", (17,0)->(17,3))
    ^^^ Token(Identifier, "neq", (17,4)->(17,7))
        ^ Token(Equal, "=", (17,8)->(17,9))
          ^ Token(Identifier, "a", (17,10)->(17,11))
            ^^ Token(NotEqualTo, "!=", (17,12)->(17,14))
               ^ Token(Identifier, "b", (17,15)->(17,16))
                 ^^ Token(Or, "||", (17,17)->(17,19))
                    ^ Token(Identifier, "a", (17,20)->(17,21))
                      ^^^ Token(NotEqualTo, "!=#", (17,22)->(17,25))
                          ^ Token(Identifier, "b", (17,26)->(17,27))
                            Token(EndOfLine, "\n", (17,27)->(17,27))
var neq_ins = a !=? b
^^^ Token(Identifier, "var", (18,0)->(18,3))
    ^^^^^^^ Token(Identifier, "neq_ins", (18,4)->(18,11))
            ^ Token(Equal, "=", (18,12)->(18,13))
              ^ Token(Identifier, "a", (18,14)->(18,15))
                ^^^ Token(NotEqualToIns, "!=?", (18,16)->(18,19))
                    ^ Token(Identifier, "b", (18,20)->(18,21))
                      Token(EndOfLine, "\n", (18,21)->(18,21))

 Token(EndOfLine, "\n", (19,0)->(19,0))
var gt = a > b || a ># b
^^^ Token(Identifier, "var", (20,0)->(20,3))
    ^^ Token(Identifier, "gt", (20,4)->(20,6))
       ^ Token(Equal, "=", (20,7)->(20,8))
         ^ Token(Identifier, "a", (20,9)->(20,10))
           ^ Token(GreaterThan, ">", (20,11)->(20,12))
             ^ Token(Identifier, "b", (20,13)->(20,14))
               ^^ Token(Or, "||", (20,15)->(20,17))
                  ^ Token(Identifier, "a", (20,18)->(20,19))
                    ^^ Token(GreaterThan, ">#", (20,20)->(20,22))
                       ^ Token(Identifier, "b", (20,23)->(20,24))
                         Token(EndOfLine, "\n", (20,24)->(20,24))
var gt_ins = a >? b
^^^ Token(Identifier, "var", (21,0)->(21,3))
    ^^^^^^ Token(Identifier, "gt_ins", (21,4)->(21,10))
           ^ Token(Equal, "=", (21,11)->(21,12))
             ^ Token(Identifier, "a", (21,13)->(21,14))
               ^^ Token(GreaterThanIns, ">?", (21,15)->(21,17))
                  ^ Token(Identifier, "b", (21,18)->(21,19))
                    Token(EndOfLine, "\n", (21,19)->(21,19))

 Token(EndOfLine, "\n", (22,0)->(22,0))
var gte = a >= b || a >=# b
^^^ Token(Identifier, "var", (23,0)->(23,3))
    ^^^ Token(Identifier, "gte", (23,4)->(23,7))
        ^ Token(Equal, "=", (23,8)->(23,9))
          ^ Token(Identifier, "a", (23,10)->(23,11))
            ^^ Token(GreaterThanOrEqual, ">=", (23,12)->(23,14))
               ^ Token(Identifier, "b", (23,15)->(23,16))
                 ^^ Token(Or, "||", (23,17)->(23,19))
                    ^ Token(Identifier, "a", (23,20)->(23,21))
                      ^^^ Token(GreaterThanOrEqual, ">=#", (23,22)->(23,25))
                          ^ Token(Identifier, "b", (23,26)->(23,27))
                            Token(EndOfLine, "\n", (23,27)->(23,27))
var gte_ins = a >=? b
^^^ Token(Identifier, "var", (24,0)->(24,3))
    ^^^^^^^ Token(Identifier, "gte_ins", (24,4)->(24,11))
            ^ Token(Equal, "=", (24,12)->(24,13))
              ^ Token(Identifier, "a", (24,14)->(24,15))
                ^^^ Token(GreaterThanOrEqualIns, ">=?", (24,16)->(24,19))
                    ^ Token(Identifier, "b", (24,20)->(24,21))
                      Token(EndOfLine, "\n", (24,21)->(24,21))

 Token(EndOfLine, "\n", (25,0)->(25,0))
var lt = a < b || a <# b
^^^ Token(Identifier, "var", (26,0)->(26,3))
    ^^ Token(Identifier, "lt", (26,4)->(26,6))
       ^ Token(Equal, "=", (26,7)->(26,8))
         ^ Token(Identifier, "a", (26,9)->(26,10))
           ^ Token(LessThan, "<", (26,11)->(26,12))
             ^ Token(Identifier, "b", (26,13)->(26,14))
               ^^ Token(Or, "||", (26,15)->(26,17))
                  ^ Token(Identifier, "a", (26,18)->(26,19))
                    ^^ Token(LessThan, "<#", (26,20)->(26,22))
                       ^ Token(Identifier, "b", (26,23)->(26,24))
                         Token(EndOfLine, "\n", (26,24)->(26,24))
var lt_ins = a <? b
^^^ Token(Identifier, "var", (27,0)->(27,3))
    ^^^^^^ Token(Identifier, "lt_ins", (27,4)->(27,10))
           ^ Token(Equal, "=", (27,11)->(27,12))
             ^ Token(Identifier, "a", (27,13)->(27,14))
               ^^ Token(LessThanIns, "<?", (27,15)->(27,17))
                  ^ Token(Identifier, "b", (27,18)->(27,19))
                    Token(EndOfLine, "\n", (27,19)->(27,19))

 Token(EndOfLine, "\n", (28,0)->(28,0))
var lte = a <= b || a <=# b
^^^ Token(Identifier, "var", (29,0)->(29,3))
    ^^^ Token(Identifier, "lte", (29,4)->(29,7))
        ^ Token(Equal, "=", (29,8)->(29,9))
          ^ Token(Identifier, "a", (29,10)->(29,11))
            ^^ Token(LessThanOrEqual, "<=", (29,12)->(29,14))
               ^ Token(Identifier, "b", (29,15)->(29,16))
                 ^^ Token(Or, "||", (29,17)->(29,19))
                    ^ Token(Identifier, "a", (29,20)->(29,21))
                      ^^^ Token(LessThanOrEqual, "<=#", (29,22)->(29,25))
                          ^ Token(Identifier, "b", (29,26)->(29,27))
                            Token(EndOfLine, "\n", (29,27)->(29,27))
var lte_ins = a <=? b
^^^ Token(Identifier, "var", (30,0)->(30,3))
    ^^^^^^^ Token(Identifier, "lte_ins", (30,4)->(30,11))
            ^ Token(Equal, "=", (30,12)->(30,13))
              ^ Token(Identifier, "a", (30,14)->(30,15))
                ^^^ Token(LessThanOrEqualIns, "<=?", (30,16)->(30,19))
                    ^ Token(Identifier, "b", (30,20)->(30,21))
                      Token(EndOfLine, "\n", (30,21)->(30,21))

 Token(EndOfLine, "\n", (31,0)->(31,0))
var reg = a =~ b || a =~# b
^^^ Token(Identifier, "var", (32,0)->(32,3))
    ^^^ Token(Identifier, "reg", (32,4)->(32,7))
        ^ Token(Equal, "=", (32,8)->(32,9))
          ^ Token(Identifier, "a", (32,10)->(32,11))
            ^^ Token(RegexpMatches, "=~", (32,12)->(32,14))
               ^ Token(Identifier, "b", (32,15)->(32,16))
                 ^^ Token(Or, "||", (32,17)->(32,19))
                    ^ Token(Identifier, "a", (32,20)->(32,21))
                      ^^^ Token(RegexpMatches, "=~#", (32,22)->(32,25))
                          ^ Token(Identifier, "b", (32,26)->(32,27))
                            Token(EndOfLine, "\n", (32,27)->(32,27))
var reg_ins = a =~? b
^^^ Token(Identifier, "var", (33,0)->(33,3))
    ^^^^^^^ Token(Identifier, "reg_ins", (33,4)->(33,11))
            ^ Token(Equal, "=", (33,12)->(33,13))
              ^ Token(Identifier, "a", (33,14)->(33,15))
                ^^^ Token(RegexpMatchesIns, "=~?", (33,16)->(33,19))
                    ^ Token(Identifier, "b", (33,20)->(33,21))
                      Token(EndOfLine, "\n", (33,21)->(33,21))

 Token(EndOfLine, "\n", (34,0)->(34,0))
var no_reg = a !~ b || a !~# b
^^^ Token(Identifier, "var", (35,0)->(35,3))
    ^^^^^^ Token(Identifier, "no_reg", (35,4)->(35,10))
           ^ Token(Equal, "=", (35,11)->(35,12))
             ^ Token(Identifier, "a", (35,13)->(35,14))
               ^^ Token(NotRegexpMatches, "!~", (35,15)->(35,17))
                  ^ Token(Identifier, "b", (35,18)->(35,19))
                    ^^ Token(Or, "||", (35,20)->(35,22))
                       ^ Token(Identifier, "a", (35,23)->(35,24))
                         ^^^ Token(NotRegexpMatches, "!~#", (35,25)->(35,28))
                             ^ Token(Identifier, "b", (35,29)->(35,30))
                               Token(EndOfLine, "\n", (35,30)->(35,30))
var no_reg_ins = a !~? b
^^^ Token(Identifier, "var", (36,0)->(36,3))
    ^^^^^^^^^^ Token(Identifier, "no_reg_ins", (36,4)->(36,14))
               ^ Token(Equal, "=", (36,15)->(36,16))
                 ^ Token(Identifier, "a", (36,17)->(36,18))
                   ^^^ Token(NotRegexpMatchesIns, "!~?", (36,19)->(36,22))
                       ^ Token(Identifier, "b", (36,23)->(36,24))
                         Token(EndOfLine, "\n", (36,24)->(36,24))

 Token(EndOfLine, "\n", (37,0)->(37,0))
var is_ = a is b || a is# b
^^^ Token(Identifier, "var", (38,0)->(38,3))
    ^^^ Token(Identifier, "is_", (38,4)->(38,7))
        ^ Token(Equal, "=", (38,8)->(38,9))
          ^ Token(Identifier, "a", (38,10)->(38,11))
            ^^ Token(Is, "is", (38,12)->(38,14))
               ^ Token(Identifier, "b", (38,15)->(38,16))
                 ^^ Token(Or, "||", (38,17)->(38,19))
                    ^ Token(Identifier, "a", (38,20)->(38,21))
                      ^^^ Token(Is, "is", (38,22)->(38,25))
                          ^ Token(Identifier, "b", (38,26)->(38,27))
                            Token(EndOfLine, "\n", (38,27)->(38,27))
var is__ins = a is? b
^^^ Token(Identifier, "var", (39,0)->(39,3))
    ^^^^^^^ Token(Identifier, "is__ins", (39,4)->(39,11))
            ^ Token(Equal, "=", (39,12)->(39,13))
              ^ Token(Identifier, "a", (39,14)->(39,15))
                ^^^ Token(IsInsensitive, "is", (39,16)->(39,19))
                    ^ Token(Identifier, "b", (39,20)->(39,21))
                      Token(EndOfLine, "\n", (39,21)->(39,21))

 Token(EndOfLine, "\n", (40,0)->(40,0))
var is_not = a isnot b || a isnot# b
^^^ Token(Identifier, "var", (41,0)->(41,3))
    ^^^^^^ Token(Identifier, "is_not", (41,4)->(41,10))
           ^ Token(Equal, "=", (41,11)->(41,12))
             ^ Token(Identifier, "a", (41,13)->(41,14))
               ^^^^^ Token(IsNot, "isnot", (41,15)->(41,20))
                     ^ Token(Identifier, "b", (41,21)->(41,22))
                       ^^ Token(Or, "||", (41,23)->(41,25))
                          ^ Token(Identifier, "a", (41,26)->(41,27))
                            ^^^^^^ Token(IsNot, "isnot", (41,28)->(41,34))
                                   ^ Token(Identifier, "b", (41,35)->(41,36))
                                     Token(EndOfLine, "\n", (41,36)->(41,36))
var is_not_ins = a isnot? b
^^^ Token(Identifier, "var", (42,0)->(42,3))
    ^^^^^^^^^^ Token(Identifier, "is_not_ins", (42,4)->(42,14))
               ^ Token(Equal, "=", (42,15)->(42,16))
                 ^ Token(Identifier, "a", (42,17)->(42,18))
                   ^^^^^^ Token(IsNotInsensitive, "isnot", (42,19)->(42,25))
                          ^ Token(Identifier, "b", (42,26)->(42,27))
                            Token(EndOfLine, "\n", (42,27)->(42,27))

